### PR TITLE
[Enhance] Use Inferencer to implement Demo

### PR DESCRIPTION
**Pull request by sunjiahao1999**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: demo/mono_det_demo.py [hunk 1]
```diff
@@ -21,50 +22,77 @@ def parse_args():
...
+        help='The interval of show (s). Demo will be blocked in showing'
+        'results, if wait_time is -1. Defaults to -1.')
     parser.add_argument(
-        '--snapshot',
+        '--no-save-vis',
         action='store_true',
-        help='whether to save online visualization results')
-    args = parser.parse_args()
-    return args
+        help='Do not save detection vis results')
```
Comment (Tai-Wang) [inline demo/mono_det_demo.py:47]: vis -> visualization

File: demo/multi_modality_demo.py [hunk 2]
```diff
@@ -1,78 +1,101 @@
...
+    parser.add_argument('infos', help='Infos file with annotations')
+    parser.add_argument('model', help='Config file')
+    parser.add_argument('weights', help='Checkpoint file')
     parser.add_argument(
         '--device', default='cuda:0', help='Device used for inference')
     parser.add_argument(
         '--cam-type',
         type=str,
-        default='CAM_FRONT',
+        default='CAM_BACK',
```
Comment (Tai-Wang) [inline demo/multi_modality_demo.py:23]: Why change this default value

File: mmdet3d/apis/inferencers/base_3d_inferencer.py [hunk 3]
```diff
@@ -258,7 +325,7 @@ def postprocess(
                 Defaults to False.
             print_result (bool): Whether to print the inference result w/o
                 visualization to the console. Defaults to False.
-            pred_out_file (str): File to save the inference results w/o
+            pred_out_dir (str): Dir to save the inference results w/o
```
Comment (Tai-Wang) [inline mmdet3d/apis/inferencers/base_3d_inferencer.py:328]: Directory

File: mmdet3d/apis/inferencers/base_3d_inferencer.py [hunk 4]
```diff
@@ -273,40 +340,85 @@ def postprocess(
               json-serializable dict containing only basic data elements such
               as strings and numbers.
         """
+        if no_save_pred is True:
```
Comment (Tai-Wang) [inline mmdet3d/apis/inferencers/base_3d_inferencer.py:343]: if no_save_pred:

File: projects/BEVFusion/demo/multi_modalit_demo.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (Tai-Wang) [inline projects/BEVFusion/demo/multi_modalit_demo.py:1]: filename: multi_modalit -> multi_modality

**Review by Tai-Wang — CHANGES_REQUESTED** [review]
Please set the default visualization option to False, because users may have trouble installing open3d on clusters but they need test the inference demo.

By the way, update the doc if necessary.

File: demo/mono_det_demo.py [hunk 6]
```diff
@@ -21,50 +22,77 @@ def parse_args():
...
+        type=str,
+        default='outputs',
+        help='Output directory of prediction results.')
     parser.add_argument(
         '--show',
         action='store_true',
-        help='show online visualization results')
+        help='Show online visualization results')
+    parser.add_argument(
+        '--wait_time',
```
Comment (ZCMax) [inline demo/mono_det_demo.py:39]: '--wait-time'

File: demo/multi_modality_demo.py [hunk 7]
```diff
@@ -1,78 +1,101 @@
...
+    parser.add_argument(
+        '--wait_time',
+        type=float,
+        default=-1,
+        help='The interval of show (s). Demo will be blocked in showing'
+        'results, if wait_time is -1. Defaults to -1.')
+    parser.add_argument(
+        '--no-save-vis',
+        action='store_true',
+        help='Do not save detection vis results')
```
Comment (ZCMax) [inline demo/multi_modality_demo.py:48]: What is the difference between detection vis and json results？ Maybe we need to clarify it.

File: demo/multi_modality_demo.py (hunk 2 above)
Comment (sunjiahao1999) [reply demo/multi_modality_demo.py:23]: Fix it.

File: demo/multi_modality_demo.py (hunk 7 above)
Comment (sunjiahao1999) [reply demo/multi_modality_demo.py:48]: visualization result is `.png` and prediction result is `.json` file

File: mmdet3d/apis/inferencers/base_3d_inferencer.py [hunk 8]
```diff
@@ -173,72 +144,90 @@ def _init_visualizer(self, cfg: ConfigType) -> Optional[Visualizer]:
...
+    def _dispatch_kwargs(self,
+                         out_dir: str = '',
+                         cam_type: str = '',
+                         **kwargs) -> Tuple[Dict, Dict, Dict, Dict]:
+        """Dispatch kwargs to preprocess(), forward(), visualize() and
+        postprocess() according to the actual demands.
+
+        Args:
+            out_dir (str): Dir to save the inference results.
+            cam_type (str): Camera type. Defaults to 'CAM2'.
```
Comment (ZCMax) [inline mmdet3d/apis/inferencers/base_3d_inferencer.py:156]: The default number is not correct.

File: mmdet3d/apis/inferencers/base_3d_inferencer.py [hunk 9]
```diff
@@ -173,72 +144,90 @@ def _init_visualizer(self, cfg: ConfigType) -> Optional[Visualizer]:
...
+        """
+        kwargs['img_out_dir'] = out_dir
+        kwargs['pred_out_dir'] = out_dir
+        if cam_type != '':
+            kwargs['cam_type_dir'] = cam_type
+        return super()._dispatch_kwargs(**kwargs)
+
     def __call__(self,
                  inputs: InputsType,
+                 batch_size=1,
```
Comment (ZCMax) [inline mmdet3d/apis/inferencers/base_3d_inferencer.py:175]: batch_size: int = 1,

File: mmdet3d/apis/inferencers/base_3d_inferencer.py [hunk 10]
```diff
@@ -173,72 +144,90 @@ def _init_visualizer(self, cfg: ConfigType) -> Optional[Visualizer]:
...
-                Defaults to 0.3.
-            img_out_dir (str): Output directory of visualization results.
-                If left as empty, no file will be saved. Defaults to ''.
-            print_result (bool): Whether to print the inference result w/o
-                visualization to the console. Defaults to False.
-            pred_out_file (str): File to save the inference results w/o
-                visualization. If left as empty, no file will be saved.
-                Defaults to ''.
-            **kwargs: Other keyword arguments passed to :meth:`preprocess`,
+            batch_size (int): Batch size. Defaults to 1.
```
Comment (ZCMax) [inline mmdet3d/apis/inferencers/base_3d_inferencer.py:184]: the parameter order should be aligned with the function.

File: mmdet3d/apis/inferencers/lidar_det3d_inferencer.py [hunk 11]
```diff
@@ -113,9 +123,10 @@ def visualize(self,
                   preds: PredType,
                   return_vis: bool = False,
                   show: bool = False,
-                  wait_time: int = 0,
+                  wait_time: int = -1,
```
Comment (ZCMax) [inline mmdet3d/apis/inferencers/lidar_det3d_inferencer.py:126]: the default value should be updated in the corresponding docstring

File: mmdet3d/apis/inferencers/lidar_det3d_inferencer.py [hunk 12]
```diff
@@ -185,3 +203,53 @@ def visualize(self,
             self.num_visualized_frames += 1
 
         return results
+
+    def visualize_preds_fromfile(self, inputs: InputsType, preds: PredType,
+                                 **kwargs) -> Union[List[np.ndarray], None]:
+        """Visualize predictions from `*.json` files.
```
Comment (ZCMax) [inline mmdet3d/apis/inferencers/lidar_det3d_inferencer.py:209]: check the doctoring parameters

File: mmdet3d/apis/inferencers/multi_modality_det3d_inferencer.py [hunk 13]
```diff
@@ -144,7 +218,9 @@ def visualize(self,
                   wait_time: int = 0,
                   draw_pred: bool = True,
                   pred_score_thr: float = 0.3,
-                  img_out_dir: str = '') -> Union[List[np.ndarray], None]:
+                  no_save_vis: bool = False,
```
Comment (ZCMax) [inline mmdet3d/apis/inferencers/multi_modality_det3d_inferencer.py:221]: update docstrings

File: projects/BEVFusion/demo/multi_modality_demo.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+def main(args):
+    # build the model from a config file and a checkpoint file
+    model = init_model(args.config, args.checkpoint, device=args.device)
+
+    # init visualizer
+    visualizer = VISUALIZERS.build(model.cfg.visualizer)
+    visualizer.dataset_meta = model.dataset_meta
+
+    # test a single image and point cloud sample
+    result, data = inference_multi_modality_detector(model, args.pcd, args.img,
```
Comment (ZCMax) [inline projects/BEVFusion/demo/multi_modality_demo.py:49]: Why do we not use the multi-modality inferencer in the demo?

Comment (sunjiahao1999) [reply projects/BEVFusion/demo/multi_modality_demo.py:49]: Because multi-modality inferencer only support one image for a smaple to inference now.

**Review by Tai-Wang — APPROVED** [review]

**Pull request merged**